### PR TITLE
Example changes for Intel PVC pipeline

### PR DESCRIPTION
**Pull request by muhammad-tanvir-1211**
This PR is a subset for #35 which includes all the changes required for writing an example for testing the Intel pipeline.

Blocked by #50 and #51, but can be reviewed independently. Also, we need to set the `sub_group` size to 16 for the output to be correct.

### CODE REVIEW

File: examples/sycl/pvc/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,34 @@
+
```
Comment (aacostadiaz) [inline examples/sycl/pvc/CMakeLists.txt:1]: same here

File: examples/CMakeLists.txt [hunk 2]
```diff
@@ -149,6 +149,7 @@ foreach(EXAMPLE
   57_hopper_grouped_gemm
   58_ada_fp8_gemm
   59_ampere_gather_scatter_conv
+  sycl
```
Comment (aacostadiaz) [inline examples/CMakeLists.txt:152]: Should we include this only if SYCL is enabled?

**Review by aacostadiaz — APPROVED** [review]
LGTM!

File: tools/util/include/cutlass/util/device_memory.h [hunk 3]
```diff
@@ -71,10 +79,14 @@ T* allocate(size_t count = 1) {
 template <typename T>
 void free(T* ptr) {
   if (ptr) {
+#if defined(CUTLASS_ENABLE_SYCL)
+    syclcompat::free(reinterpret_cast<void*>(ptr));
```
Comment (AD2605) [inline tools/util/include/cutlass/util/device_memory.h:83]:
```suggestion
    if(ptr) {
        syclcompat::free(reinterpret_cast<void*>(ptr));
    }
```
nullptr check before freeing

File: tools/util/include/cutlass/util/device_memory.h [hunk 4]
```diff
@@ -140,12 +156,16 @@ class DeviceAllocation {
   /// Delete functor for CUDA device memory
   struct deleter {
     void operator()(T* ptr) {
+#ifdef CUTLASS_ENABLE_SYCL
+      syclcompat::free(reinterpret_cast<void*>(ptr));
```
Comment (AD2605) [inline tools/util/include/cutlass/util/device_memory.h:160]: nullptr check here as well

File: tools/util/include/cutlass/util/reference/device/tensor_foreach.h [hunk 5]
```diff
@@ -65,12 +69,19 @@ struct TensorForEach {
...
       // single thread and reduces the impact of initialization overhead.
       block_size = (block_size < 128 ? block_size : 128);
+#endif
     }
 
     dim3 grid(grid_size, 1, 1);
     dim3 block(block_size, 1, 1);
 
+#if defined(CUTLASS_ENABLE_SYCL)
+    const auto sycl_block = syclcompat::dim3(block.x, block.y, block.z);
```
[the remark this answers is not in the document]
Comment (AD2605) [reply tools/util/include/cutlass/util/reference/device/tensor_foreach.h:79]: Same comment for other places as well

File: tools/util/include/cutlass/util/device_memory.h (hunk 3 above)
Comment (muhammad-tanvir-1211) [reply tools/util/include/cutlass/util/device_memory.h:83]: There is a check for the nullptr before the #ifdef statement.

File: tools/util/include/cutlass/util/device_memory.h (hunk 4 above)
Comment (muhammad-tanvir-1211) [reply tools/util/include/cutlass/util/device_memory.h:160]: There is a nullptr check for this in the file https://github.com/codeplaysoftware/cutlass-fork/blob/8a456718f944bdb6497b114a336c5d8138579234/include/cutlass/platform/platform.h#L842

File: tools/util/include/cutlass/util/device_memory.h [hunk 6]
```diff
@@ -58,11 +58,19 @@ T* allocate(size_t count = 1) {
 
   bytes = count * sizeof(T);
 
+#if defined(CUTLASS_ENABLE_SYCL)
+  if (bytes > 0) {
+    ptr = reinterpret_cast<T*>(syclcompat::malloc(bytes));
```
[the remark this answers is not in the document]
Comment (AD2605) [reply tools/util/include/cutlass/util/device_memory.h:63]: Ideally it should be aligned by the underlying driver.
Resolving this , and we can report it if it ever poses a problem

**Review by AD2605 — APPROVED** [review]

**Pull request merged**